### ISSUE TITLE
 Problem with integration with keycloak

### ISSUE THREAD

**Issue by 0st0ne** [zmartzone/lua-resty-openidc#514]
###### Environment

- lua-resty-openidc version (e.g. 1.7.6)
- Keycloak 23.0.7
- OpenResty 1.25.3.1

###### Expected behaviour

I have the following openresty configuration:

##I added these dictionaries based on recommendations from previous threads on this issue
lua_shared_dict discovery 1m;
lua_shared_dict jwks 1m;
lua_shared_dict introspection 10m;
lua_shared_dict sessions 10m;

##For correct operation of ssl
lua_ssl_verify_depth 5;
lua_ssl_trusted_certificate /etc/ssl/certs/ca-certificates.pem;

server {
 listen 443 ssl;

 ssl_certificate    /etc/ssl/certs/domain.crt;
 ssl_certificate_key    /etc/ssl/private/domain.key;

 access_log /home/portal/logs/log.log;
 error_log /home/portal/logs/log_error.log info;

 resolver 8.8.8.8;

 fastcgi_read_timeout 600;
 client_header_timeout 30;
##I added these session secret based on recommendations from previous threads on this issue
 set $session_secret 1234567890;

 location / {
      access_by_lua_block {
         local opts = {
           redirect_uri = "https://domain.com/callback",
           discovery = "https://sso.domain.com/realms/ditust/.well-known/openid-configuration",
           client_id = "test_app",
           client_secret = "123123123132132132132132131323211",
           ssl_verify = "no",
        }

        -- call authenticate for OpenID Connect user authentication
        local res, err,target, session = require("resty.openidc").authenticate(opts)
        session:close()
        if err then
              ngx.status = 403
              ngx.say(err)
              ngx.exit(ngx.HTTP_FORBIDDEN)
        end

        ngx.req.set_header("X-USER", res.id_token.sub)
    }

    proxy_pass https://test.domain.com/
  }
 }

Steps to reproduce:
1. go to site https://domain.com/;
2. redirects me to KC;
3. successful authentication;
4. redirects back to https://domain.com/callback with authentication headers;
5. got an error request to the redirect_uri path but there's no session state found.

More details:
When I go to my site https://domain.com/
I get a redirect to keycloak.
After successful authentication, I receive a reverse redirect to my site with all the headers necessary for authentication, but it returns a 403 error.
This is due to the fact that the authenticate(opts) method could not return res and lua will throw a 403 error.

After my research, I found out that the problem is that openresty does not remember the session. the session.create() method, when used again, does not restore the previous session, but creates a new one.
This can also be observed from the error in the logs:

openidc.lua:1511: authenticate(): request to the redirect_uri path but there's no session state found

After researching a lot of articles, I think that this configuration should work correctly in this form

If you have any ideas or suggestions on how to help, I would be grateful

thank you in advance!

**Comment by haglo** [comment]
I have exactly the same error after i updated my Windows to Windows 11 Version 23H2. Before all worked perfectly.

**Comment by bodewig** [comment]
`request to the redirect_uri path but there's no session state found` almost always means your browser did not send the session cookie lua-resty-session expects to be present. This can happen for a number of reasons and your browser's developer tools may help.

The most common case is the Cookie's `SameSite` setting being to strict for your OIDC setup - for example if you use the "form_post" response_mode then `SameSite` must be set to `None` as even `Lax` would be too strict. For the "normal" redirect based response mode `SameSite=Lax` should work. `SameSite=Strict`will not work with any OIDC login flow at all. See https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Set-Cookie#samesitesamesite-value for the attribute. See https://github.com/bungle/lua-resty-session/tree/v3.10?tab=readme-ov-file#string-sessioncookiesamesite for the way to configure the session cookie's SameSite value.

**Comment by haglo** [comment]
Still not working. This is my Configuration:
`
worker_processes  1;

events {
  worker_connections 128;
}

http {
  lua_package_path '~/lua/?.lua;;';
  resolver 127.0.0.11 ipv6=off;
  lua_ssl_trusted_certificate /etc/ssl/certs/ca-certificates.crt;
  lua_ssl_verify_depth 5;
  lua_shared_dict discovery 1m;
  lua_shared_dict jwks 1m;

  server {
	server_name localhost;
  listen 0.0.0.0:80;
	listen [::]:80;


    location /auth {
        proxy_pass https://<realm>/;
        proxy_set_header Host $http_host;
        proxy_pass_request_headers on;
    }
	

    location / {
      set $session_cookie_samesite off;
      access_by_lua_block {
        local opts = {
          redirect_uri_path = "/redirect_uri",
          discovery = "https://<domain>/realms/<realm>/.well-known/openid-configuration",
          client_id = "<client_id",
          client_secret = "<secret>",
          scope = "openid email",
          access_token_expires_leeway = 30,
          accept_none_alg = false,
          accept_unsupported_alg = false,
          renew_access_token_on_expiry = true,
          logout_path = "/logout",
          redirect_after_logout_uri = "https://<domain>/realms/<realm>/protocol/openid-connect/logout",
          redirect_after_logout_with_id_token_hint = false, 
          session_contents = {access_token=true}
        }
        local res, err, target, session = require("resty.openidc").authenticate(opts)
        session:close()

        if err then
          ngx.status = 403
          ngx.say(err)
          ngx.exit(ngx.HTTP_FORBIDDEN)
        end
        ngx.req.set_header("Authorization", "Bearer " .. res.access_token)
        ngx.req.set_header("X-User", res.id_token.sub)
        ngx.header['Authorization'] = "Bearer " .. res.access_token
      }

      proxy_pass http://host.docker.internal:8000/;

      expires    0;
      add_header Cache-Control private;
      proxy_set_header Host $http_host;
      proxy_http_version 1.1;
      proxy_redirect off;
      proxy_buffering off;
      proxy_set_header Upgrade $http_upgrade;
      proxy_set_header Connection "upgrade";
    }
  }
}
`

The strange thing is that it worked until yesterday, before i made the update of Widnows 11

**Comment by bodewig** [comment]
`off` is not going to help you at all, it means "don't set the SameSite value at all when setting the Cookie" which leaves you at the default of your browser - which is `Lax` for all modern browsers.

Most likely your Windows update has given you a different Browser with changed (default) settings.

The most reliable source for debugging the problem are the developer tools of your browser. Does you browser see the Cooke when lua-resty-openidc redirects your browser to your OIDC provider? Does it reject it for any reason? Does it send it back when returning? If not, does it tell you why not?

If it does send back the Cookie, then SameSite is not the issue and the problem is a different one.

**Comment by haglo** [comment]
<img width="1911" alt="iceScrum" src="https://github.com/zmartzone/lua-resty-openidc/assets/513155/b4e67245-46ef-4702-9888-de630f7bc372">
Attached i send you the ScreenShot, now with
` set $session_cookie_samesite Lax;`

**Comment by bodewig** [comment]
Sending screenshots is not going to help, I'm afraid. You need to look at the individual requests (the last two in your screenshot) in your dev tools and see whether the cookies are present and if it is not present in the second to see why it is not included.This may or may not get logged to your browser's console. Please understand I'm trying to help you to help yourself here as the problem is with your setup.

In order to rule out SameSite, you should set the value to None and verify this is what the Set-Cookie header says when redirecting to Keycloak. You may also want to check whether any kind of 3rd party cookie policy is present that would reject 3rd party cookies regardless of SameSite=None.

**Comment by haglo** [comment]
I think it is a problem of keycloak

**Comment by 0st0ne** [comment]
Could there be a problem on the keyCloak server side? maybe there is a parameter that prevents cookies from being returned?

**Comment by bodewig** [comment]
This is not how Cookies work. Keycloak has no way to influence the cookies your browser sends to your relying party server.

**Comment by haglo** [comment]
Where is then the place to search for this problem?

**Comment by 0st0ne** [comment]
I decided to try authentication through different combinations of browsers and systems.
Everywhere there are problems that cookies are not saved in the browser.
![sso_test](https://github.com/zmartzone/lua-resty-openidc/assets/60350554/e4b02257-5cf2-429e-b256-55aefb2a9ad5)

**Comment by lexiconzero** [comment]
You might have done what I just spent 5 hours troubleshooting..

[lua-resty-session](https://github.com/bungle/lua-resty-session) needs to be fixed at version 3.10 or lower. If you've pulled the latest by defult, these kinds of errors will pop up

**Comment by 0st0ne** [comment]
Are you saying that you were able to get a working state of the system by running it on a version lower than 3.10?

**Comment by lexiconzero** [comment]
That's right - pinning lua-resty-session to version 3.10 corrected this issue

**Comment by 0st0ne** [comment]
@lexiconzero 
It didn't work for me(
I have the following error in the logs 

lua entry thread aborted: runtime error: /usr/local/openresty/lualib/resty/openidc.lua:965: bad argument #1 to 'match' (string expected, got nil)

**Comment by lexiconzero** [comment]
Here's my code for the location I'm using in openresty. Ive also set a session secret above in the server block.

`location / {
          access_by_lua_block { 
            local opts = {
            redirect_uri = "/redirect_uri",
            accept_none_alg = false,
            renew_access_token_on_expiry = true,
            discovery = "http://keycloak:8080/realms/test/.well-known/openid-configuration",
            client_id = "nginx",
            client_secret = "xxx",
            logout_path = "/logout",
            redirect_after_logout_with_id_token_hint = true,
            redirect_uri_scheme = "https",
            accept_unsupported_alg = false,
            renew_access_token_on_expiry = true,
            revoke_tokens_on_logout = true,
            ssl_verify = "no",
            session_contents = {id_token=true}
       }
   
       local oidc = require("resty.openidc")
       
       oidc.set_logging(nil, { DEBUG = ngx.INFO })

       local res, err = oidc.authenticate(opts)
       if err then
         ngx.status = 403
         ngx.exit(ngx.HTTP_FORBIDDEN)
       end
       }
          proxy_set_header Host host;
          proxy_set_header Upgrade $http_upgrade;
          proxy_set_header Connection "upgrade";
          proxy_pass_request_headers on;
          proxy_cache_bypass $cookie_session;
          proxy_no_cache $cookie_session;
          proxy_pass      http://ipaddress:8123;

      }`

Here's how I load this into the podman container it runs in via a compose file

`openresty:
hostname: hostname
restart: unless-stopped
image: docker.io/bitnami/openresty:latest
volumes:
  - /home/user/openrestyconf/openresty-homeassist.conf:/opt/bitnami/openresty/nginx/conf/server_blocks/openresty-homeassist.conf:Z
  - /home/user/openrestyconf/openresty-keycloak.conf:/opt/bitnami/openresty/nginx/conf/server_blocks/openresty-keycloak.conf:Z
  - /home/user/openrestyconf/lua-opm-files.sh:/docker-entrypoint-initdb.d/lua-opm-files.sh:Z
ports:
  - 3002:3002
  - 3003:3003
depends_on:
  - keycloak_postgress
  - keycloak


#Lua-opm-files.sh
#!/bin/bash
opm install ledgetech/lua-resty-http
opm install bungle/lua-resty-session=3.10
opm install cdbattags/lua-resty-jwt
opm install zmartzone/lua-resty-openidc

`

**Comment by samuelmonsempessenthorus** [comment]
Any news about that ?

**Comment by samuelmonsempessenthorus** [comment]
What is the issue and how can I help to resolve this ?

**Comment by zRich** [comment]
my web app runs on a forked domain at https://console.szqrinfo.com/ 
my keycloak ruans on https://auth.szqrinfo.com/
both enable https, with the same cert.

my keycloak configration is most likely the example config for oauth

`local KEYCLOAK_INTERNAL_ENDPOINT = "http://auth.szqrinfo.com"
    local KEYCLOAK_EXTERNAL_ENDPOINT = "https://auth.szqrinfo.com"
    local KEYCLOAK_LOGOUT_REDIRECT_URI = "/"
    local KEYCLOAK_REALM = "jwt-demo"
    local KEYCLOAK_CLIENT = "console"
    local KEYCLOAK_SECRET = "j4r4i4sZes8TzDglATDRLLNaLWQ18jPd"

    local opts = {
            redirect_uri = "https://console.szqrinfo.com/",
            --redirect_uri = "/redirect_uri",

            -- The discovery endpoint of the OP. Enable to get the URI of all endpoints (Token, introspection, logout...)
            discovery = "https://auth.szqrinfo.com/realms/jwt-demo/.well-known/openid-configuration",

            -- Access to OP Token endpoint requires an authentication. Several authentication modes are supported:
            --token_endpoint_auth_method = ["client_secret_basic"|"client_secret_post"|"private_key_jwt"|"client_secret_jwt"],
            -- o If token_endpoint_auth_method is set to "client_secret_basic", "client_secret_post", or "client_secret_jwt", authentication to Token endpoint is using client_id and client_secret
        
            client_id = "platform",
            client_secret = "BNne1fe8ccZvsKp5dr8VLtdwrij8oipd",
            
            ssl_verify = "no",
            -- Connection keepalive with the OP can be enabled ("yes") or disabled ("no").
            --keepalive = "no",

            --response_mode=form_post
            scope = "openid email profile",
            -- Refresh the users id_token after 900 seconds without requiring re-authentication
            --refresh_session_interval = 900,
            --iat_slack = 600,
            --redirect_uri_scheme = "https",
            --logout_path = "/logout",
            --redirect_after_logout_uri = "/",
            -- Where should the user be redirected after logout from the RP. This option overides any end_session_endpoint that the OP may have provided in the discovery response.
            --redirect_after_logout_with_id_token_hint = true,
            -- Whether the redirection after logout should include the id token as an hint (if available). This option is used only if redirect_after_logout_uri is set.
            --post_logout_redirect_uri = "https://www.zmartzone.eu/logoutSuccessful",
            -- Where should the user be redirected after logout from the OP. This option overides any post_logout_redirect_uri that the OP may have provided in the discovery response.
            --accept_none_alg = false
            
            --accept_unsupported_alg = true
            -- if you want to reject tokens signed using an algorithm
            -- not supported by lua-resty-jwt set this to false. If
            -- you leave it unset or set it to true, the token signature will not be
            -- verified when an unsupported algorithm is used.

            --renew_access_token_on_expiry = true
            -- whether this plugin shall try to silently renew the access token once it is expired if a refresh token is available.
            -- if it fails to renew the token, the user will be redirected to the authorization endpoint.
            --access_token_expires_in = 3600
            -- Default lifetime in seconds of the access_token if no expires_in attribute is present in the token endpoint response.

            --access_token_expires_leeway = 0
            --  Expiration leeway for access_token renewal. If this is set, renewal will happen access_token_expires_leeway seconds before the token expiration. This avoids errors in case the access_token just expires when arriving to the OAuth Resource Server.

            force_reauthorize = true,
            -- When force_reauthorize is set to true the authorization flow will be executed even if a token has been cached already
            session_contents = {id_token=true, access_token=true},
            -- Whitelist of session content to enable. This can be used to reduce the session size.
            -- When not set everything will be included in the session.
            -- Available are:
            -- id_token, enc_id_token, user, access_token (includes refresh token)

            -- You can specify timeouts for connect/send/read as a single number (setting all timeouts) or as a table. Values are in milliseconds
            -- timeout = 1000
            -- timeout = { connect = 500, send = 1000, read = 1000 }

            --use_nonce = false
            -- By default the authorization request includes the
            -- nonce paramter. You can use this option to disable it
            -- which may be necessary when talking to a broken OpenID
            -- Connect provider that ignores the paramter as the
            -- id_token will be rejected otherwise.

            --revoke_tokens_on_logout = false
        }`

my configration for session:

sessions on login page
![Image](https://github.com/user-attachments/assets/3e0244c2-3d46-42b6-9766-dedcd70fc3f4)

sessoins after login page

![Image](https://github.com/user-attachments/assets/2ddaeb1d-ecd0-4d52-ad24-79a171a1e913)

the sessions of both https://console.szqrinfo.com/ and  https://auth.szqrinfo.com/ are identical

but keycloak doesn't redirect to back the original page

![Image](https://github.com/user-attachments/assets/47fdec9d-23ff-4303-9e7a-be6fe2d35afb)

can not figure out a solution so far

**Comment by asyd** [comment]
We have the same issue with 1.8 while it's working fine with 1.7.6-3 (with the same configuration)

**Comment by darkglade** [comment]
> You might have done what I just spent 5 hours troubleshooting..
> 
> [lua-resty-session](https://github.com/bungle/lua-resty-session) needs to be fixed at version 3.10 or lower. If you've pulled the latest by defult, these kinds of errors will pop up

This fixed the issue for me too...

EDIT: I see there's some updates in the README, I'll have to digest that.

**Comment by darkglade** [comment]
Yeah, no joy, I've added the session_opts as described in the README but it STILL fails with the current version of lua-resty-session (4.1.5) same config works fine with 3.10...

```
local opts = {
  redirect_uri_path = "/redirect_uri",
  discovery = "<DISCO URL>",
  token_signing_alg_values_expected = "ES256",
  client_id = "<CLIENT ID>",
  client_secret = "<CLIENT SECRET>",
  redirect_after_logout_uri = "<IDP HOME URL>",
  redirect_after_logout_with_id_token_hint = false,
  use_pkce = true
}

local session_opts = {
  secret = "<SECRET>",
  cookie_http_only = true,
  cookie_secure = true,
  cookie_same_site = "Lax",
  remember = true
}

...

local res, err = require("resty.openidc").authenticate(opts, nil, nil, session_opts)
```

Still getting the "authenticate(): request to the redirect_uri path but there's no session state found" error in the logs.

I can see the session cookie on the request to redirect_uri in devtools.

**Comment by darkglade** [comment]
OK, this is odd, same config (even without the session_opts) works fine using the docker alpine-fat image, but running direct on an Ubuntu host I'm having this issue.

Ah, OK, might be a difference between the opm package and the luarocks package... opm installed 1.7.5 and luarocks installed 1.8.0 and the dependencies are a bit different.

That resolved the issue for me, removed all the opm packages and installed lua-resty-openidc using luarocks (which annoyingly required the installation of a bunch of dev tools I could've done without) and it works now (even with the current version of lua-resty-session) seems there's some breaking change in lua-resty-session v4+ which breaks lua-resty-openidc 1.7.5 so I suppose the question is, why is the opm version not 1.8.0? (I DID install the zmartzone/lua-resty-openidc version not one of the others available from opm)
